common/src/main/java/org/apache/celeborn/common/network/sasl/SaslClientBootstrap.java

```diff
@@ -0,0 +1,120 @@
+
+import static org.apache.celeborn.common.network.sasl.SaslUtils.*;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeoutException;
+
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+import io.netty.channel.Channel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.celeborn.common.network.client.TransportClient;
+import org.apache.celeborn.common.network.client.TransportClientBootstrap;
+import org.apache.celeborn.common.network.protocol.TransportMessage;
+import org.apache.celeborn.common.network.util.TransportConf;
+import org.apache.celeborn.common.protocol.MessageType;
+import org.apache.celeborn.common.protocol.PbAuthType;
+import org.apache.celeborn.common.protocol.PbSaslRequest;
+import org.apache.celeborn.common.util.JavaUtils;
+
+/**
+ * Bootstraps a {@link TransportClient} by performing SASL authentication on the connection. The
+ * server should be setup with a {@code SaslRpcHandler} with matching keys for the given appId.
+ */
+public class SaslClientBootstrap implements TransportClientBootstrap {
+  private static final Logger logger = LoggerFactory.getLogger(SaslClientBootstrap.class);
+
+  private final TransportConf conf;
+  private final String appId;
+  private final SaslCredentials saslCredentials;
+
+  /**
+   * Creates a new SaslClientBootstrap. When this is used for authenticating a client connection,
+   * the user id in the saslCredentials should be the same as the appId.
+   *
+   * @param conf transport conf
+   * @param appId application id
+   * @param saslCredentials sasl userId and password.
+   */
+  public SaslClientBootstrap(TransportConf conf, String appId, SaslCredentials saslCredentials) {
+    this.conf = conf;
+    this.appId = appId;
+    this.saslCredentials = Preconditions.checkNotNull(saslCredentials);
+  }
+
+  /**
+   * Performs SASL authentication by sending a token, and then proceeding with the SASL
+   * challenge-response tokens until we either successfully authenticate or throw an exception due
+   * to mismatch.
+   */
+  @Override
+  public void doBootstrap(TransportClient client, Channel channel) {
+    // TODO: Hardcoding the SASL mechanism to DIGEST-MD5 for Connection Authentication. This
+    // should be configurable in the future.
+    CelebornSaslClient saslClient =
+        new CelebornSaslClient(
+            DIGEST,
+            DEFAULT_SASL_CLIENT_PROPS,
+            new CelebornSaslClient.ClientCallbackHandler(
+                saslCredentials.getUserId(), saslCredentials.getPassword()));
+    try {
+      byte[] payload = saslClient.firstToken();
+
+      while (!saslClient.isComplete()) {
+        TransportMessage msg =
+            new TransportMessage(
+                MessageType.SASL_REQUEST,
+                PbSaslRequest.newBuilder()
+                    .setMethod(DIGEST)
+                    .setAuthType(PbAuthType.CONNECTION_AUTH)
```

Comment:
We need to set the method and auth type only in the initial request, right ?
Rest is just an exchange of challenge and response.